lib/mongoid/touchable.rb

```diff
@@ -22,46 +22,73 @@ module InstanceMethods
+
+        touches = __gather_touch_updates(Time.configured.now, field)
+        _root.send(:persist_atomic_operations, '$set' => touches) if touches.present?
+
+        __run_touch_callbacks_from_root
+        true
+      end
+
+      # Recursively sets touchable fields on the current document and each of its
+      # parents, including the root node. Returns the combined atomic $set
+      # operations to be performed on the root document.
+      #
+      # @param [ Time ] now The timestamp used for synchronizing the touched time.
+      # @param [ Symbol ] field The name of an additional field to update.
+      #
+      # @return [ Hash<String, Time> ] The touch operations to perform as an atomic $set.
+      #
+      # @api private
+      def __gather_touch_updates(now, field = nil)
         field = database_field_name(field)
-        write_attribute(:updated_at, current) if respond_to?("updated_at=")
-        write_attribute(field, current) if field
-
-        # If the document being touched is embedded, touch its parents
-        # all the way through the composition hierarchy to the root object,
-        # because when an embedded document is changed the write is actually
-        # performed by the composition root. See MONGOID-3468.
-        if _parent
-          # This will persist updated_at on this document as well as parents.
-          # TODO support passing the field name to the parent's touch method;
-          # I believe it should be read out of
-          # _association.inverse_association.options but inverse_association
-          # seems to not always/ever be set here. See MONGOID-5014.
-          _parent.touch
-
-          if field
-            # If we are told to also touch a field, perform a separate write
-            # for that field. See MONGOID-5136.
-            # In theory we should combine the writes, which would require
-            # passing the fields to be updated to the parents - MONGOID-5142.
-            sets = set_field_atomic_updates(field)
-            selector = atomic_selector
-            _root.collection.find(selector).update_one(positionally(selector, sets), session: _session)
-          end
-        else
-          # If the current document is not embedded, it is composition root
-          # and we need to persist the write here.
-          touches = touch_atomic_updates(field)
-          unless touches["$set"].blank?
-            selector = atomic_selector
-            _root.collection.find(selector).update_one(positionally(selector, touches), session: _session)
-          end
-        end
+        write_attribute(:updated_at, now) if respond_to?("updated_at=")
+        write_attribute(field, now) if field
 
-        # Callbacks are invoked on the composition root first and on the
-        # leaf-most embedded document last.
+        touches = __extract_touches_from_atomic_sets(field) || {}
+        touches.merge!(_parent.__gather_touch_updates(now) || {}) if __touchable_parent?
```

Comment:
So I'm tracing through this code and this feels a little repetitive, can you explain why this is necessary? I'll explain... 

At the deepest child (take the building > floor > sofa example), the `__gather_touch_updates` call returns:
```
{ "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
Now one level up from that (the floor level): 
```
{ "floors.0.updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
And at the top (building) level: 
```
{ "updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.updated_at" => 2022-10-03 15:12:53 UTC, "floors.0.sofas.0.updated_at" => 2022-10-03 15:12:53 UTC }
```
This, I think, results from the fact that the atomic updates of a child are included in its parent (does this hold always?) 

Would a more efficient solution be to find the greatest touchable ancestor and ``__extract_touches_from_atomic_sets`` once from there?

Reply:
If I remember correctly, the reason for doing this is that, not only do we want to get the document updates (which are applied to root), we also want to mutate the objects in-memory to reflect the touch. Hence all the `write_attribute` calls. Does this make sense?

Reply:
you can do all that without calling `__extract_touches_from_atomic_sets` on each "generation", no?

Reply:
My brain is out of mana on this issue. I suggest we merge as-is and do the refactor as a follow-up.

You're probably right that it can be made more efficient in Ruby, however, the first thing this PR is trying to achieve is to make the database calls more efficient (which it achieves).

Reply:
I'm not quite ready to approve this, if you don't mind just thinking out this last point with me.... I decided to see how difficult this would be to fix, and I discovered an interesting decision you made (I'm not sure if this was accidentally or purposefully), but I would like to bring it up.... 

I noticed that in the recursive call to `_gather_touch_updates` the field is not passed to the parent as well. Now, is this done on purpose, since the given field should only be touched on the deepest level, whereas on its ancestors the `updated_at` field should be touched? 

If this is true, then it makes sense to call `_extract_touches_from_atomic_sets` on each level.

Reply:
However I noticed one other thing which is _not_ taken into account, and I would be okay with leaving that as a followup, but MONGOID-5508 is not taken into account here. The recursive call here would have to check the association for a touch field and pass it in.

Reply:
Verify the above comments for me and then I should be okay to approve.

Reply:
> Is this done on purpose, since the given field should only be touched on the deepest level, whereas on its ancestors the updated_at field should be touched.

The intent is that `child.touch(:field_a)` would touch `:field_a` and `:updated` on the child, and then *only* touch `:updated_at` (not `:field_a`) on the parent. Supposing I also have:

```
class Child
  embedded_in :parent, touch: :field_b
```

Then I need to touch `:field_b` and `:updated_at` on the parent, regardless of which field I touch on the child.

Does that make sense?

Reply:
As we've noted elsewhere, there are more refactors needed especially for persistence-related auto-touching, so I expect we'll be refactoring all this again anyway.

Reply:
yes it does make sense.... My latter point was that this: 
>  Then I need to touch :field_b and :updated_at on the parent, regardless of which field I touch on the child.

is not being accomplished in this PR. I'm okay with that, and I'm willing for that to be done in MONGOID-5508.